### PR TITLE
Clustering summary

### PR DESCRIPTION
**Pull request by Avgor46**

### CODE REVIEW

File: src/bin/casr-cli.rs [hunk 1]
```diff
@@ -36,8 +44,15 @@ fn main() -> Result<()> {
     let matches = App::new("casr-cli")
         .author("Andrey Fedotov <fedotoff@ispras.ru>, Alexey Vishnyakov <vishnya@ispras.ru>, Georgy Savidov <avgor46@ispras.ru>")
         .version("2.3.0")
-        .about("App provides text-based user interface to view CASR reports")
+        .about("App provides text-based user interface to view CASR reports and print summary for clustering")
         .term_width(90)
+        .arg(
```
Comment (SweetVishnya) [inline src/bin/casr-cli.rs:49]: We may remove logs

File: src/bin/casr-cli.rs [hunk 2]
```diff
@@ -629,3 +679,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+        for report_path in fs::read_dir(cluster).unwrap().flatten() {
+            let input = report_path.path();
+            if input.is_file()
+                && input.extension().is_some()
+                && input.extension().unwrap() == "casrep"
+            {
+                // input == .*/crash-hash.gdb.casrep
+                let mut report = input.canonicalize().unwrap().with_extension("");
+                // report == .*/crash-hash.gdb
+                if report.extension().is_some() && report.extension().unwrap() == "gdb" {
```
Comment (SweetVishnya) [inline src/bin/casr-cli.rs:737]: We should support arbitrary extensions. Not only gdb.

File: src/bin/casr-cli.rs [hunk 3]
```diff
@@ -629,3 +679,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+        // Hash files
+        let mut filestems: HashSet<PathBuf> = HashSet::new();
+        for report_path in fs::read_dir(cluster).unwrap().flatten() {
+            let input = report_path.path();
+            if input.is_file()
+                && input.extension().is_some()
+                && input.extension().unwrap() == "casrep"
+            {
+                // input == .*/crash-hash.gdb.casrep
+                let mut report = input.canonicalize().unwrap().with_extension("");
```
Comment (SweetVishnya) [inline src/bin/casr-cli.rs:735]: This is actually input, and input above is actually report. Maybe rename?

File: src/bin/casr-cli.rs (hunk 2 above)
Comment (SweetVishnya) [reply src/bin/casr-cli.rs:737]: Moreover, there may be more than 2 reports describing single crash.

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: src/bin/casr-cli.rs (hunk 2 above)
Comment (Avgor46) [reply src/bin/casr-cli.rs:737]: "." can appear anywhere in the filename, support for arbitrary extensions is difficult

File: src/bin/casr-cli.rs [hunk 4]
```diff
@@ -36,8 +44,15 @@ fn main() -> Result<()> {
     let matches = App::new("casr-cli")
         .author("Andrey Fedotov <fedotoff@ispras.ru>, Alexey Vishnyakov <vishnya@ispras.ru>, Georgy Savidov <avgor46@ispras.ru>")
         .version("2.3.0")
-        .about("App provides text-based user interface to view CASR reports")
+        .about("App provides text-based user interface to view CASR reports and print summary for clustering")
```
[the remark this answers is not in the document]
Comment (Avgor46) [reply src/bin/casr-cli.rs]: done

File: src/bin/casr-cli.rs [hunk 5]
```diff
@@ -36,8 +44,15 @@ fn main() -> Result<()> {
...
         .term_width(90)
```
[the remark this answers is not in the document]
Comment (Avgor46) [reply src/bin/casr-cli.rs:47]: done

File: src/bin/casr-cli.rs (hunk 1 above)
Comment (Avgor46) [reply src/bin/casr-cli.rs:48]: done

File: src/bin/casr-cli.rs [hunk 6]
```diff
@@ -629,3 +679,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+///
+/// # Arguments
+///
+/// * 'report' - path to report
+///
+/// * 'extension' - casrep extension
+///
+/// * 'line' - mutable argument for crash line
+///
+/// * 'desc' - mutable argument for short description
```
[the remark this answers is not in the document]
Comment (Avgor46) [reply src/bin/casr-cli.rs]: done

File: src/bin/casr-cli.rs (hunk 3 above)
Comment (Avgor46) [reply src/bin/casr-cli.rs]: done

File: src/bin/casr-cli.rs [hunk 7]
```diff
@@ -629,3 +660,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+        let mut cluster_classes: HashMap<String, i32> = HashMap::new();
+        // Hash files
+        let mut filestems: HashSet<PathBuf> = HashSet::new();
+        for report_path in fs::read_dir(cluster).unwrap().flatten() {
+            let report = report_path.path();
+            if report.is_file()
+                && report.extension().is_some()
+                && report.extension().unwrap() == "casrep"
+            {
+                // report == .*/crash-hash.gdb.casrep
```
Comment (anfedotoff) [inline src/bin/casr-cli.rs:715]: maybe just crash.gdb.casrep? It is not always libFuzzer format. It is just an example

File: src/bin/casr-cli.rs [hunk 8]
```diff
@@ -629,3 +660,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+/// # Arguments
+///
+/// * 'report' - path to report
+///
+/// * 'extension' - casrep extension
+///
+/// # Return value
+///
+/// String - summary
+///
```
Comment (anfedotoff) [inline src/bin/casr-cli.rs:828]: Remove this line

File: src/bin/casr-cli.rs [hunk 9]
```diff
@@ -629,3 +660,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+fn process_report(report: &str, extension: &str) -> Result<(String, String, String)> {
```
Comment (anfedotoff) [inline src/bin/casr-cli.rs:829]: You don't use Result, error mesage. Use Option then.

File: src/bin/casr-cli.rs [hunk 10]
```diff
@@ -629,3 +660,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+            let severity = jreport["CrashSeverity"]["Type"]
+                .as_str()
+                .unwrap()
+                .to_string();
+            let crashline = if let Some(crashline) = jreport.get("CrashLine") {
+                let crashline_str = crashline.as_str().unwrap();
+                if !crashline_str.trim().is_empty() {
+                    crashline_str.to_string()
+                } else {
+                    String::new()
```
Comment (anfedotoff) [inline src/bin/casr-cli.rs:846]: you could put `report` here

File: src/bin/casr-cli.rs [hunk 11]
```diff
@@ -629,3 +660,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+                    "{}: {}: {}: {}",
+                    extension,
+                    match severity.as_str() {
+                        "EXPLOITABLE" => "EXPLOITABLE".red(),
+                        "PROBABLY_EXPLOITABLE" => "PROBABLY_EXPLOITABLE".yellow(),
+                        "NOT_EXPLOITABLE" => "NOT_EXPLOITABLE".white(),
+                        &_ => "UNDEFINED".white(),
+                    },
+                    desc,
+                    if crashline.is_empty() {
```
Comment (anfedotoff) [inline src/bin/casr-cli.rs:862]: just use crashline

File: src/bin/casr-cli.rs [hunk 12]
```diff
@@ -629,3 +660,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+///
+/// # Arguments
+///
+/// * 'report' - path to report
+///
+/// * 'extension' - casrep extension
+///
+/// # Return value
+///
+/// String - summary
```
Comment (anfedotoff) [inline src/bin/casr-cli.rs:827]: need more information

File: src/bin/casr-cli.rs [hunk 13]
```diff
@@ -629,3 +660,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+///
+fn process_report(report: &str, extension: &str) -> Result<(String, String, String)> {
+    if let Ok(file) = fs::File::open(report) {
```
Comment (anfedotoff) [inline src/bin/casr-cli.rs:830]: try this pattern: https://rust-lang.github.io/rfcs/3137-let-else.html

**Review by anfedotoff — CHANGES_REQUESTED** [review]

File: src/bin/casr-cli.rs (hunk 10 above)
Comment (Avgor46) [reply src/bin/casr-cli.rs:846]: Then if it was not possible to get the crashline from reports, hashes will be different for all reports

File: docs/usage.md [hunk 14]
```diff
@@ -207,13 +207,13 @@ Similar CASR reports are inside one cluster.
 
 ## casr-cli
 
-App provides text-based user interface to view CASR reports
+App provides text-based user interface to view CASR reports and print summary of clusters
```
[the remark this answers is not in the document]
Comment (Avgor46) [reply docs/usage.md]: done

File: src/bin/casr-cli.rs [hunk 15]
```diff
@@ -629,3 +660,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
         Action::Mouse(_) => None,
     }
 }
+
+/// Function prints summary after casr work
```
[the remark this answers is not in the document]
Comment (Avgor46) [reply src/bin/casr-cli.rs]: done

File: src/bin/casr-cli.rs [hunk 16]
```diff
@@ -629,3 +660,215 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+        && fs::read_dir(&casr_dir)
+            .unwrap()
+            .filter(|res| res.is_ok())
+            .map(|res| res.unwrap().path())
+            .any(|e| e.extension().is_some() && e.extension().unwrap() == "casrep")
+    {
+        clusters.push((casr_dir, 0));
+    }
+
+    for cl_path in clusters {
```
[the remark this answers is not in the document]
Comment (Avgor46) [reply src/bin/casr-cli.rs]: yep

File: src/bin/casr-cli.rs (hunk 9 above)
Comment (Avgor46) [reply src/bin/casr-cli.rs]: done

File: src/bin/casr-cli.rs (hunk 13 above)
Comment (Avgor46) [reply src/bin/casr-cli.rs]: done

File: src/bin/casr-cli.rs [hunk 17]
```diff
@@ -629,3 +660,217 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+                    result.push(crash);
+                }
+
+                let mut hash = String::new();
+                hash.push_str(san_desc.as_str());
+                hash.push_str(san_line.as_str());
+                hash.push_str(casr_gdb_desc.as_str());
+                hash.push_str(casr_gdb_line.as_str());
+
+                let san_cnt = cluster_classes.get(&san_desc).unwrap_or(&0) + 1;
```
Comment (anfedotoff) [inline src/bin/casr-cli.rs:772]: What will be if `san_desc` is empty? We have only .gdb.casreps

Comment (anfedotoff) [reply src/bin/casr-cli.rs:772]: Maybe we should check this case?

File: src/bin/casr-cli.rs [hunk 18]
```diff
@@ -629,3 +660,217 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
     }
 }
+
+/// Print common statistic all over reports in directory
+///
+/// # Arguments
+///
+/// * 'dir' - directory with reports
+///
+fn print_summary(dir: PathBuf) -> Result<()> {
```
Comment (anfedotoff) [inline src/bin/casr-cli.rs:670]: We don't return any errors. Maybe remove Result?

File: src/bin/casr-cli.rs [hunk 19]
```diff
@@ -629,3 +660,217 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+                let value = if let Some(res) = cluster_hash.get(&hash) {
+                    (res.0.clone(), res.1 + 1)
+                } else {
+                    (result, 1)
+                };
+                cluster_hash.insert(hash, value);
+            }
+        }
+        for info in cluster_hash.values() {
+            info!("{}: {}", "Crash".green(), info.0.last().unwrap());
```
Comment (anfedotoff) [inline src/bin/casr-cli.rs:788]: Maybe some comments here? for each log line?

File: src/bin/casr-cli.rs [hunk 20]
```diff
@@ -629,3 +660,217 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
...
+            info!("  {}", info.0[0]);
+            if info.0.len() == 3 {
+                info!("  {}", info.0[1]);
+            }
+            info!("  Similar crashes: {}", info.1);
+        }
+        let mut classes = String::new();
+        cluster_classes.iter().for_each(|(class, number)| {
+            classes.push_str(format!(" {}: {}", class, number).as_str());
```
Comment (anfedotoff) [inline src/bin/casr-cli.rs:797]: Need to rebase on master and fix  clippy

**Review by anfedotoff — CHANGES_REQUESTED** [review]

File: src/bin/casr-cli.rs [hunk 21]
```diff
@@ -824,51 +824,53 @@ fn print_summary(casr_dir: PathBuf) -> Result<()> {
...
+/// 2 String - crash description
+/// 3 String - crashline (path:line:column)
+fn process_report(report: &str, extension: &str) -> Option<(String, String, String)> {
+    let Ok(file) = fs::File::open(report) else {
+        return None;
+    };
+    let Ok(jreport) = serde_json::from_reader(BufReader::new(file)) else {
+        return None;
+    };
+    let jreport: Value = jreport;
```
[the remark this answers is not in the document]
Comment (Avgor46) [reply src/bin/casr-cli.rs]: reworked

File: src/bin/casr-cli.rs (hunk 17 above)
Comment (Avgor46) [reply src/bin/casr-cli.rs]: done

File: src/bin/casr-cli.rs (hunk 18 above)
Comment (Avgor46) [reply src/bin/casr-cli.rs]: done

File: src/bin/casr-cli.rs (hunk 19 above)
Comment (Avgor46) [reply src/bin/casr-cli.rs:791]: done

File: Cargo.toml [hunk 22]
```diff
@@ -35,6 +35,8 @@ rayon = "~1.5"
 num_cpus = "~1.13"
 linux-personality = "~1.0"
 thiserror = "~1.0"
+colored = "~2.0"
+fern = { version = "~0.6", features = ["colored"] }
```
Comment (SweetVishnya) [inline Cargo.toml:39]: Do we still need fern?

File: docs/usage.md [hunk 23]
```diff
@@ -225,13 +225,14 @@ Frames that match these regular expressions will be not considered during analys
 
 ## casr-cli
 
-App provides text-based user interface to view CASR reports
+App provides text-based user interface to view CASR reports and print common statistic all
```
Comment (SweetVishnya) [inline docs/usage.md:228]: and print joint statistics for multiple reports

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: docs/usage.md (hunk 23 above)
Comment (anfedotoff) [reply docs/usage.md:228]: maybe for all? Multiple != all

Comment (SweetVishnya) [reply docs/usage.md:228]: > maybe for all? Multiple != all

ok

File: Cargo.toml (hunk 22 above)
Comment (Avgor46) [reply Cargo.toml:39]: removed

File: docs/usage.md (hunk 23 above)
Comment (Avgor46) [reply docs/usage.md]: done

**Review by SweetVishnya — APPROVED** [review]

**Review by anfedotoff — APPROVED** [review]

**Pull request merged**